OSDOCS-9735: add customize subcommand iscsi procedures (#77322)
openshift/openshift-docs: modules/installation-user-infra-machines-advanced-customizing-live-iscsi-ibft.adoc

// Module included in the following assemblies
//
// * installing/installing_bare_metal/installing-bare-metal.adoc
// * installing/installing_bare_metal/installing-restricted-networks-bare-metal.adoc
// * installing_bare_metal/installing-bare-metal-network-customizations.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-user-infra-machines-advanced-customizing-live-{boot}-iscsi-ibft_{context}"]
= Customizing a live install {boot-media} for an iSCSI boot device with iBFT

You can set the iSCSI target and initiator values for automatic mounting, booting and configuration using a customized version of the live {op-system} image.

.Prerequisites
. You have an iSCSI target you want to install {op-system} on.

> . Optional: you have multipathed your iSCSI target.

Author: @GroceryBoyJr it's got some precedence both in OCP docs and RHEL docs, so i think it's fine

.Procedure

. Download the `coreos-installer` binary from the link:https://mirror.openshift.com/pub/openshift-v4/clients/coreos-installer/latest/[`coreos-installer` image mirror] page.

ifeval::["{boot-media}" == "ISO image"]
. Retrieve the {op-system} ISO image from the link:https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/latest/[{op-system} image mirror] page and run the following command to customize the ISO image with the following information:
+
[source,text]
----
$ coreos-installer iso customize \
    --pre-install mount-iscsi.sh \ <1>
    --post-install unmount-iscsi.sh \ <2>
    --dest-device /dev/mapper/mpatha \ <3>
    --dest-ignition config.ign \ <4>
    --dest-karg-append rd.iscsi.firmware=1 \ <5>
    --dest-karg-append rd.multipath=default \ <6>
    -o custom.iso rhcos-<version>-live.x86_64.iso
----

> <1> The script that gets run before installation. It should contain the `iscsiadm` commands for mounting the iSCSI target and any commands enabling multipathing.

Reviewer: Overall looks good to me, one small question, not the blocker, does `mount-iscsi.sh` is like the `iscsiadm` command in https://github.com/openshift/openshift-docs/pull/77275 ?

Author: yeah, pretty much! 😄 i think links between the two procedures is a good idea, i'll take a look at adding some post-GA

<2> The script that gets run after installation. It should contain the command `iscsiadm --mode node --logout=all`.

> <3> The path of a single multipathed device. If there are multiple multipath devices connected, or to be explicit, you can use the World Wide Name (WWN) symlink available in `/dev/disk/by-path`.

Reviewer 2: ```suggestion
<3> The path to the device. If using multipath, the multipath device (e.g. `/dev/mapper/mpatha`). If there are multiple multipath devices connected, or to be explicit, you can use the World Wide Name (WWN) symlink available in `/dev/disk/by-path`.
```

Reviewer 2: Similarly for PXE.

> <4> The ignition configuration for the destination system.

Reviewer 2: ```suggestion
<4> The Ignition configuration for the destination system.
```

+
[source,text]
----
$ coreos-installer pxe customize \
    --pre-install mount-iscsi.sh \ <1>
    --post-install unmount-iscsi.sh \ <2>
    --dest-device /dev/mapper/mpatha \ <3>
    --dest-ignition config.ign \ <4>
    --dest-karg-append rd.iscsi.firmware=1 \ <5>
    --dest-karg-append rd.multipath=default \ <6>
    -o custom.img rhcos-<version>-live-initramfs.x86_64.img
----
<1> The script that gets run before installation. It should contain the `iscsiadm` commands for mounting the iSCSI target.
<2> The script that gets run after installation. It should contain the command `iscsiadm --mode node --logout=all`.
<3> The path of a single multipathed device. If there are multiple multipath devices connected, or to be explicit, you can use the World Wide Name (WWN) symlink available in `/dev/disk/by-path`.
<4> The ignition configuration for the destination system.
<5> The iSCSI parameter is read from the BIOS firmware.
<6> Optional: include this parameter if you are enabling multipathing.
endif::[]
+

> For more information about the iSCSI options supported by `dracut`, see the link:https://www.man7.org/linux/man-pages/man7/dracut.cmdline.7.html[`dracut.cmdline` manual page].

Reviewer 3: I love how you set this up. It's clean and easy to read. But I have a question: are the scripts you are talking about provided by us at Red Hat or do they need to be written by the user? (and, please pardon me if I missed something along the way and should have known this)

Author: The `dracut` commands i believe are inherited from RHEL

Author: @GroceryBoyJr ohh i see what you mean. you're referring to the `pre-install` and `post-install` scripts mentioned. they need to be created by the user so kind of outside the scope of what we document. however, they need to include the specific things mentioned + whatever else they need for their infrastructure
